From f9985cc43c9e9424242c7408c20d2d2379f8e788 Mon Sep 17 00:00:00 2001
From: Wei Li <liwei391@huawei.com>
Date: Wed, 14 Sep 2022 18:00:34 +0800
Subject: [PATCH 0210/1076] [Huawei] kallsyms: export
 kallsyms_lookup_name_wrap() and kallsyms_on_each_symbol_wrap()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: NA

--------------------------------

kallsyms_lookup_name() and kallsyms_on_each_symbol() were unexported by
commit 0bd476e6c671 ("kallsyms: unexport kallsyms_lookup_name() and
kallsyms_on_each_symbol()"), while some drivers still need to use them.

Add two wrap functions controlled by option 'use_kallsyms_lookup_symbol'
and export them again for limited usage.

Signed-off-by: Wei Li <liwei391@huawei.com>
---
 include/linux/kallsyms.h |  4 ++++
 kernel/kallsyms.c        | 29 +++++++++++++++++++++++++++++
 2 files changed, 33 insertions(+)

diff --git a/include/linux/kallsyms.h b/include/linux/kallsyms.h
index 481273f0c72d..16c8b913fdd4 100644
--- a/include/linux/kallsyms.h
+++ b/include/linux/kallsyms.h
@@ -74,11 +74,15 @@ static inline void *dereference_symbol_descriptor(void *ptr)
 #ifdef CONFIG_KALLSYMS
 /* Lookup the address for a symbol. Returns 0 if not found. */
 unsigned long kallsyms_lookup_name(const char *name);
+unsigned long kallsyms_lookup_name_wrap(const char *name);
 
 /* Call a function on each kallsyms symbol in the core kernel */
 int kallsyms_on_each_symbol(int (*fn)(void *, const char *, struct module *,
 				      unsigned long),
 			    void *data);
+int kallsyms_on_each_symbol_wrap(int (*fn)(void *, const char *, struct module *,
+				      unsigned long),
+			    void *data);
 
 extern int kallsyms_lookup_size_offset(unsigned long addr,
 				  unsigned long *symbolsize,
diff --git a/kernel/kallsyms.c b/kernel/kallsyms.c
index c6738525fe11..cf04dcb2f7fd 100644
--- a/kernel/kallsyms.c
+++ b/kernel/kallsyms.c
@@ -196,6 +196,35 @@ int kallsyms_on_each_symbol(int (*fn)(void *, const char *, struct module *,
 	return module_kallsyms_on_each_symbol(fn, data);
 }
 
+bool use_kallsyms_lookup_symbol __ro_after_init;
+
+static int __init kallsyms_lookup_symbol_setup(char *str)
+{
+	use_kallsyms_lookup_symbol = true;
+	return 1;
+}
+__setup("use_kallsyms_lookup_symbol", kallsyms_lookup_symbol_setup);
+
+unsigned long kallsyms_lookup_name_wrap(const char *name)
+{
+	if (!use_kallsyms_lookup_symbol)
+		return 0;
+
+	return kallsyms_lookup_name(name);
+}
+EXPORT_SYMBOL_GPL(kallsyms_lookup_name_wrap);
+
+int kallsyms_on_each_symbol_wrap(int (*fn)(void *, const char *, struct module *,
+				      unsigned long),
+			    void *data)
+{
+	if (!use_kallsyms_lookup_symbol)
+		return -EPERM;
+
+	return kallsyms_on_each_symbol(fn, data);
+}
+EXPORT_SYMBOL_GPL(kallsyms_on_each_symbol_wrap);
+
 static unsigned long get_symbol_pos(unsigned long addr,
 				    unsigned long *symbolsize,
 				    unsigned long *offset)
-- 
2.17.1

